File: src/main/java/StockData.java
/**
 * Plain data object holding all the metrics retrieved for a single stock.
 *
 * Fields are populated by {@link YahooFinanceFetcher#fetch(String)} and may be
 * {@code Double.NaN} when the API does not supply a value (e.g. a stock that
 * pays no dividend will have {@code dividendYield == NaN}).  Numeric fields that
 * default to {@code 0.0} instead of NaN (price, change, volume, marketCap) are
 * always present in a valid API response, so 0 is treated as "not available" for
 * display purposes only.
 */
public class StockData {

    // ---- Identity ------------------------------------------------------------

    /** Ticker symbol in upper-case, e.g. "AAPL". */
    public String symbol;

    /** Human-readable company name, e.g. "Apple Inc." */
    public String companyName;

    /** Exchange the stock is listed on, e.g. "NASDAQ" or "NYSE". */
    public String exchange;

    /** ISO 4217 currency code for all price fields, e.g. "USD". */
    public String currency;

    // ---- Price ---------------------------------------------------------------

    /** Most recent trade price. */
    public double currentPrice;

    /** Absolute price change vs. previous close (can be negative). */
    public double priceChange;

    /** Percentage change vs. previous close, e.g. 2.35 means +2.35 %. */
    public double priceChangePercent;

    // ---- Market data ---------------------------------------------------------

    /** Total market capitalisation in the stock's currency. */
    public double marketCap;

    /** Most recent session's trading volume in shares. */
    public long tradingVolume;

    /** 3-month average daily trading volume in shares. */
    public long averageDailyVolume;

    // ---- Valuation ratios ----------------------------------------------------

    /** Trailing twelve-month price-to-earnings ratio. */
    public double peRatio;

    /** Forward (estimated next-12-month) price-to-earnings ratio. */
    public double forwardPE;

    /** Trailing twelve-month earnings per share. */
    public double earningsPerShare;

    /** Price-to-book ratio. */
    public double priceToBook;

    // ---- Risk & income -------------------------------------------------------

    /**
     * Beta — measures volatility relative to the broader market.
     * A value above 1.0 means more volatile than the market.
     */
    public double beta;

    /**
     * Annual dividend yield stored as a <em>decimal fraction</em>
     * (e.g. 0.0235 means 2.35 %).  Multiply by 100 before displaying as a
     * percentage.  {@code NaN} when the stock pays no dividend.
     */
    public double dividendYield;

    // ---- Price range & moving averages ---------------------------------------

    /** Highest closing price over the trailing 52 weeks. */
    public double fiftyTwoWeekHigh;

    /** Lowest closing price over the trailing 52 weeks. */
    public double fiftyTwoWeekLow;

    /** 50-day simple moving average of closing prices. */
    public double fiftyDayMovingAverage;

    /** 200-day simple moving average of closing prices. */
    public double twoHundredDayMovingAverage;

    public StockData() {}
}

File: MobilePortStock-iOS/MobilePortStock/Models/StockData.swift
import Foundation

/// All metrics retrieved for a single stock quote.
///
/// Fields are populated by ``YahooFinanceService/fetch(_:)`` and use Swift
/// optionals (`nil`) where the Java original used `Double.NaN` to signal
/// missing data.
struct StockData: Codable, Identifiable, Hashable {

    // MARK: - Identity

    /// Ticker symbol in upper-case, e.g. "AAPL".
    let symbol: String

    /// Human-readable company name, e.g. "Apple Inc."
    let companyName: String

    /// Exchange the stock is listed on, e.g. "NASDAQ" or "NYSE".
    let exchange: String

    /// ISO 4217 currency code for all price fields, e.g. "USD".
    let currency: String

    // MARK: - Price

    /// Most recent trade price.
    let currentPrice: Double

    /// Absolute price change vs. previous close (can be negative).
    let priceChange: Double

    /// Percentage change vs. previous close, e.g. 2.35 means +2.35%.
    let priceChangePercent: Double

    // MARK: - Market data

    /// Total market capitalisation in the stock's currency.
    let marketCap: Double

    /// Most recent session's trading volume in shares.
    let tradingVolume: Int

    /// 3-month average daily trading volume in shares.
    let averageDailyVolume: Int

    // MARK: - Valuation ratios

    /// Trailing twelve-month price-to-earnings ratio.
    let peRatio: Double?

    /// Forward (estimated next-12-month) price-to-earnings ratio.
    let forwardPE: Double?

    /// Trailing twelve-month earnings per share.
    let earningsPerShare: Double?

    /// Price-to-book ratio.
    let priceToBook: Double?

    // MARK: - Risk & income

    /// Beta — measures volatility relative to the broader market.
    let beta: Double?

    /// Annual dividend yield as a decimal fraction (0.0235 = 2.35%).
    /// `nil` when the stock pays no dividend.
    let dividendYield: Double?

    // MARK: - Price range & moving averages

    /// Highest closing price over the trailing 52 weeks.
    let fiftyTwoWeekHigh: Double?

    /// Lowest closing price over the trailing 52 weeks.
    let fiftyTwoWeekLow: Double?

    /// 50-day simple moving average of closing prices.
    let fiftyDayMovingAverage: Double?

    /// 200-day simple moving average of closing prices.
    let twoHundredDayMovingAverage: Double?

    // MARK: - Identifiable

    var id: String { symbol }
}
